### PR TITLE
Feature/reset password

### PR DESCRIPTION
**Pull request by kobyfogel**
Reset a password by mail confirmation.
In the case of a user have forgotten his password, he can reset it by verifying his mail address.
The email verification is done by sending a mail with a  link to a reset-password page in our site, containing a jwt-token with the user's unique credentials, and 15 minutes expiration time. Upon clicking that link, decryption of that token will verify and authenticate the user. Then he will be presented with a reset-password form

### CODE REVIEW

File: app/routers/reset_password.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    return templates.TemplateResponse("forgot_password.html", {
+            "request": request,
+            "message": 'Please check your credentials'
+        })
+
+
+@router.get("/reset-password")
+async def reset_password_form(
+        request: Request, token: Optional[str] = ""
+        ) -> templates:
```
Comment (PureDreamer) [inline app/routers/reset_password.py]: why not RedirectResponse?

File: app/templates/forgot_password.html [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{% extends "base.html" %}
+{% block content %}
+    <div class="container">
+        <h5>Please verify your Username and Email:</h4>
+        {% if message %}
+            <div class="alert" style="color: red;">{{ message }}</div>
```
Comment (PureDreamer) [inline app/templates/forgot_password.html]: no inline styling

File: app/templates/reset_password_mail.html [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<h1>Hi, {{ recipient }}!</h1>
+<h2>You received this email from {{ site_name }} because you asked to reset your password.</h2>
```
Comment (PureDreamer) [inline app/templates/reset_password_mail.html]: wouldnt it be better to write down the name? its not supposed to change

File: app/templates/reset_password_mail.html [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+<p>To continue with resetting your password, please click the confirmation <a href={{ link }}>link</a></p>
+<p>This confirmation link will expired in 15 minutes</p>
```
Comment (PureDreamer) [inline app/templates/reset_password_mail.html:4]: very nice work all of this!

File: tests/test_reset_password.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    assert res.status_code == HTTP_302_FOUND
+
+
+def test_reset_password_expired_token(session, security_test_client):
+    security_test_client.post('/register', data=REGISTER_DETAIL)
+    user = ForgotPassword(**FORGOT_PASSWORD_DETAILS)
+    token = create_jwt_token(user, jwt_min_exp=-1)
+    link = f"/reset-password?token={token}"
+    res = security_test_client.post(link, data=RESET_PASSWORD_DETAILS)
+    # assert b'Your token has expired' in res.content
```
Comment (PureDreamer) [inline tests/test_reset_password.py]: delete this please

**Review by PureDreamer — COMMENTED** [review]
Very nice work!

File: app/internal/security/schema.py [hunk 6]
```diff
@@ -15,3 +15,57 @@ class LoginUser(BaseModel):
...
+    details sending a token via email
+    """
+    username: str
+    email: str
+    user_id: Optional[str] = None
+    token: Optional[str] = None
+    is_manager: Optional[bool] = False
+
+    class Config:
+        orm_mode = True
```
Comment (ivarshav) [inline app/internal/security/schema.py:42]: Consider adding example - https://fastapi.tiangolo.com/tutorial/schema-extra-example/

File: app/internal/security/schema.py [hunk 7]
```diff
@@ -15,3 +15,57 @@ class LoginUser(BaseModel):
...
+
+    @validator('username')
```
Comment (ivarshav) [inline app/internal/security/schema.py]: cool!

File: app/static/style.css [hunk 8]
```diff
@@ -103,6 +103,12 @@ p {
     padding-left: 12.5rem;
 }
 
+.forgot-password{
```
Comment (ivarshav) [inline app/static/style.css]: Add a space between d and { :)

File: app/routers/reset_password.py (hunk 1 above)
Comment (kobyfogel) [reply app/routers/reset_password.py]: This function handles 'GET' method request, and renders the empty html form.

File: app/templates/reset_password_mail.html (hunk 4 above)
Comment (kobyfogel) [reply app/templates/reset_password_mail.html:4]: Thanks!

File: app/internal/security/schema.py (hunk 6 above)
Comment (kobyfogel) [reply app/internal/security/schema.py:42]: Can you please elaborate how will it contribute to the code functionality?

**Comment by kobyfogel** [comment]
WIP! I Forgot to add at the commit name.
Few things will be updated in this PR once feature\login (current_user) is merged.

File: app/internal/security/schema.py (hunk 6 above)
Comment (ivarshav) [reply app/internal/security/schema.py:42]: For easier debugging, and for readability for other developers (for examples - in #docs views it's more readable)

Comment (kobyfogel) [reply app/internal/security/schema.py:42]: For my opinion, debugging is good enough just by using pydantic the way i did, and it's readable enough for any developer who knows how to work with pydantic models. Also, i added docstring. Thank You for your review.

**Comment by yammesicka** [comment]
Still WIP?

**Comment by kobyfogel** [comment]
> Still WIP?
Not anymore

File: app/config.py.example [hunk 9]
```diff
@@ -63,6 +63,8 @@ email_conf = ConnectionConfig(
 JWT_KEY = "JWT_KEY_PLACEHOLDER"
 JWT_ALGORITHM = "HS256"
 JWT_MIN_EXP = 60 * 24 * 7
+CALENDAR_RESET_PASSWORD_PAGE = "calendar.pythonic.guru/reset-password"
```
Comment (yammesicka) [inline app/config.py.example]: Prefer relative

File: app/internal/email.py [hunk 10]
```diff
@@ -177,3 +182,27 @@ def verify_email_pattern(email: str) -> bool:
...
+    :param user: ForgotPassword schema.
+        Contains user's email address, jwt verifying token.
+    :param background_tasks: (BackgroundTasks): Function from fastapi that lets
+            you apply tasks in the background.
+    returns True
+    """
+    template = templates.get_template("reset_password_mail.html")
+    html = template.render(
+         recipient=user.username,
+         link=f"{CALENDAR_RESET_PASSWORD_PAGE}?token={user.token}",
```
Comment (yammesicka) [inline app/internal/email.py]: Maybe url_for?

File: app/static/style.css [hunk 11]
```diff
@@ -103,6 +103,18 @@ p {
...
+    line-height: 0;
+    color: rgb(188, 7, 194);
+    padding-left: 8rem;
+}
+
+.red-massage {
+  color: red;
+}
+.subtitle {
+    font-size: 20px;
```
Comment (yammesicka) [inline app/static/style.css]: Use `rem`

**Review by yammesicka — COMMENTED** [review]
Great work! I've added few small insights. Please take a look

File: app/internal/email.py (hunk 10 above)
Comment (kobyfogel) [reply app/internal/email.py]: Is that even possible? I'm not sending a request here, its a link that is attached to an a-tag at the sent email template

File: app/config.py.example (hunk 9 above)
Comment (kobyfogel) [reply app/config.py.example]: This link is being used in a sent email, and its purpose is to redirect the user back to the website

Comment (yammesicka) [reply app/config.py.example]: So please separate the domain from the reset-password. That way the setting will be reusable for future features.

(Also, please note we already have a `DOMAIN` variable)

File: app/internal/email.py (hunk 10 above)
Comment (yammesicka) [reply app/internal/email.py]: Have you tried? :)

File: app/internal/security/ouath2.py [hunk 12]
```diff
@@ -30,19 +39,38 @@ def verify_password(plain_password: str, hashed_password: str) -> bool:
...
+async def authenticate_user_by_email(
+        db: Session, user: schema.ForgotPassword,
+        email: bool = False) -> Union[schema.ForgotPassword, bool]:
+    """
+    Verifying database record by username.
+    Comparing given email to database record,
+    """
+    db_user = await User.get_by_username(db=db, username=user.username)
+    if not db_user:
+        return False
```
Comment (yammesicka) [inline app/internal/security/ouath2.py:60]: Prefer to raise an exception

File: app/internal/security/ouath2.py [hunk 13]
```diff
@@ -30,19 +39,38 @@ def verify_password(plain_password: str, hashed_password: str) -> bool:
     return pwd_context.verify(plain_password, hashed_password)
 
 
+async def authenticate_user_by_email(
```
Comment (yammesicka) [inline app/internal/security/ouath2.py]: That's not what the function does. `is_email_compatible_to_username`

File: app/routers/reset_password.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    form_dict = dict(form)
+    try:
+        # validating form data by creating pydantic schema object
+        user = ForgotPassword(**form_dict)
+    except ValidationError:
+        user = False
+    if user:
+        user = await authenticate_user_by_email(db, user)
+        if user:
+            user.token = create_jwt_token(user, jwt_min_exp=15)
```
Comment (yammesicka) [inline app/routers/reset_password.py]: So the user is authenticated by just entering any valid mail address?

**Review by yammesicka — CHANGES_REQUESTED** [review]
Looks like great start!

Currently, It's hard to understand where the authentication of the user happens.

File: app/routers/reset_password.py (hunk 14 above)
Comment (kobyfogel) [reply app/routers/reset_password.py]: user is authenticated here, by entering his username and email. both details are being checked against those on the db,
if they both correct, a token will be generated. this token is not for logging in. its for verifying that the email address is really in use by the user

File: app/internal/security/ouath2.py (hunk 12 above)
Comment (kobyfogel) [reply app/internal/security/ouath2.py:60]: raise an exception to the user? what i'm doing here is re-rendering the form, asking him to check his credentials

**Comment by kobyfogel** [comment]
> Looks like great start!
> 
> Currently, It's hard to understand where the authentication of the user happens.
> Why are we creating a JWT token for someone who doesn't logged in yet?

The jwt token is sent by mail to the user. It's purpose is to verify that the user is using the mail address he registered with.
About the DOMAIN variable, this is what we currently got at config:
image

About using url_for with the link:
I tried to use it, it doesn't work outside the website. When i click the link in the email, i get an error

File: app/config.py.example [hunk 15]
```diff
@@ -16,7 +16,7 @@ class Settings(BaseSettings):
 
 
 # GENERAL
-DOMAIN = 'Our-Domain'
+DOMAIN = 'https://calendar.pythonic.guru/'
```
Comment (yammesicka) [inline app/config.py.example]: Please don't prefill it :)

File: app/config.py.example [hunk 16]
```diff
@@ -63,6 +63,8 @@ email_conf = ConnectionConfig(
 JWT_KEY = "JWT_KEY_PLACEHOLDER"
 JWT_ALGORITHM = "HS256"
 JWT_MIN_EXP = 60 * 24 * 7
+CALENDAR_RESET_PASSWORD_PAGE = "/reset-password"
```
Comment (yammesicka) [inline app/config.py.example]: That part shouldn't be in the configuration

File: app/internal/email.py [hunk 17]
```diff
@@ -177,3 +182,28 @@ def verify_email_pattern(email: str) -> bool:
...
+    :param background_tasks: (BackgroundTasks): Function from fastapi that lets
+            you apply tasks in the background.
+    returns True
+    """
+    params = f"?email_verification_token={user.email_verification_token}"
+    template = templates.get_template("reset_password_mail.html")
+    html = template.render(
+         recipient=user.username,
+         link=f"{DOMAIN}{CALENDAR_RESET_PASSWORD_PAGE}{params}",
+         email=user.email)
```
Comment (yammesicka) [inline app/internal/email.py]: Use git precommit hooks

File: app/internal/email.py [hunk 18]
```diff
@@ -8,8 +8,13 @@
 from sqlalchemy.orm.session import Session
 
 from app.config import (CALENDAR_HOME_PAGE, CALENDAR_REGISTRATION_PAGE,
-                        CALENDAR_SITE_NAME, email_conf, templates)
+                        CALENDAR_RESET_PASSWORD_PAGE, CALENDAR_SITE_NAME,
+                        DOMAIN, email_conf
+                        )
```
Comment (yammesicka) [inline app/internal/email.py]: Please use git precommit hooks to reformat

File: app/routers/reset_password.py [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    The contains the verafiction jwt token.
+    """
+    form = await request.form()
+    form_dict = dict(form)
+    try:
+        # validating form data by creating pydantic schema object
+        user = ForgotPassword(**form_dict)
+    except ValidationError:
+        user = False
+    if user:
```
Comment (yammesicka) [inline app/routers/reset_password.py]: Flip and use guard here (if not user: return, unindent the rest)

File: app/routers/reset_password.py [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    then sending email to the user with a reset password route link.
+    The contains the verafiction jwt token.
+    """
+    form = await request.form()
+    form_dict = dict(form)
+    try:
+        # validating form data by creating pydantic schema object
+        user = ForgotPassword(**form_dict)
+    except ValidationError:
+        user = False
```
Comment (yammesicka) [inline app/routers/reset_password.py]: `None` is more appropriate here

File: app/routers/reset_password.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+        if user:
+            user.email_verification_token = create_jwt_token(
+                user, jwt_min_exp=15)
+            await send_reset_password_mail(user, background_tasks)
+            return templates.TemplateResponse("forgot_password.html", {
+                "request": request,
+                "message": "Email for reseting password was sent"})
+    return templates.TemplateResponse("forgot_password.html", {
+            "request": request,
+            "message": 'Please check your credentials'
```
Comment (yammesicka) [inline app/routers/reset_password.py]: Use either `"` or `'`.
Add trailing comma.
Use git precommit hooks to fix both.

File: app/routers/reset_password.py [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    Rendering reset password form get method.
+    Validating jwt token is supplied with request.
+    """
+    if email_verification_token:
+        return templates.TemplateResponse("reset_password.html", {
+            "request": request,
+        })
+    message = "?message=Verification token is missing"
+    return RedirectResponse(f"/login{message}")
```
Comment (yammesicka) [inline app/routers/reset_password.py]: Use `url_for`

File: app/routers/reset_password.py [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+        user = ResetPassword(**form_dict)
+    except ValueError:
+        validated = False
+    if not validated:
+        return templates.TemplateResponse("reset_password.html", {
+            "request": request,
+            "message": 'Please check your credentials'})
+    await update_password(db, jwt_username, user.password)
+    return RedirectResponse(
+        url="/login?message=Success reset password",
```
Comment (yammesicka) [inline app/routers/reset_password.py]: Use url_for for all links in the project.

File: app/routers/reset_password.py (hunk 14 above)
Comment (yammesicka) [reply app/routers/reset_password.py]: But what holds the user from just taking the token and logging in using it?

Comment (kobyfogel) [reply app/routers/reset_password.py]: Logging in to the system does not uses token, but password.
Getting to a protected route requires that the token will be attached to an http cookie. This one is not.
At protected routes with dependencies, this token wouldn't pass the pydantic validations, as it does not carry a user id. In any case, i have added another layer of security at the dependencies, that this token cannot pass.
Routes dependencies have been tested with this token, attached to an HTTP cookie.
Accessing into protected routes is not possible by using this token.

File: app/templates/reset_password.html [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+                <div class="col-auto">
+                    <span id="PasswordHelpInline" class="form-text">
+                        Must be between 3 to 20 characters.
+                    </span>
+                </div>
+            </div>
+            <div class="row">
+                <label for="confirm_password" class="col-sm-2 col-form-label">Confirm Password</label>
+                <div class="col-auto">
+                    <input type="text" class="form-control" name="confirm_password" id="confirm_password" placeholder="Required"><br>
```
Comment (Gonzom) [inline app/templates/reset_password.html]: multi-words id names should like classes be separated with a dash and not an underscore. So "confirm-password".

File: app/templates/reset_password.html [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+                </div>
+                <div class="col-auto">
+                    <span id="ConfirmPasswordHelpInline" class="form-text">
```
Comment (Gonzom) [inline app/templates/reset_password.html]: multi-words id names should like classes be separated with a dash and not pascal case.

File: app/templates/reset_password.html (hunk 24 above)
Comment (kobyfogel) [reply app/templates/reset_password.html:29]: This Id is being sent directly to pydantic class from the route. Changing that means adding few totally unnecessary code lines in the route, just to change that comma back to a syntax that pydantic class can work with. I prefer this solution

File: app/templates/reset_password.html (hunk 25 above)
Comment (kobyfogel) [reply app/templates/reset_password.html:32]: This Id is being sent directly to pydantic class from the route. Changing that means adding few totally unnecessary code lines in the route, just to change that comma back to a syntax that pydantic class can work with. I prefer this solution

File: app/routers/reset_password.py (hunk 19 above)
Comment (yammesicka) [reply app/routers/reset_password.py]: Bump

File: app/routers/reset_password.py [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+            "request": request,
+            "message": "Email for reseting password was sent",
+        },
+    )
+
+
+@router.get("/reset-password")
+async def reset_password_form(
+    request: Request,
+    email_verification_token: Optional[str] = "",
```
Comment (yammesicka) [inline app/routers/reset_password.py:82]: That's not a good idea for many reasons.

1. GET requests saved in the history, in the server's logs and in the browser. Sending the JWT using get could help 3rd party malicious player to get the token and impersonate someone without him/her knowing it.
2. The same goes for sending the JWT via unencrypted mail.

File: app/templates/reset_password.html [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+                </div>
+            </div>
+            <div class="row">
+                <label for="confirm-password" class="col-sm-2 col-form-label">Confirm Password</label>
+                <div class="col-auto">
+                    <input type="text" class="form-control" name="confirm-password" id="confirm-password"
+                        placeholder="Required"><br>
+                </div>
+                <div class="col-auto">
+                    <span id="ConfirmPasswordHelpInline" class="form-text">
```
Comment (yammesicka) [inline app/templates/reset_password.html]: confirm-password-help-inline

File: app/templates/reset_password.html [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+                        autofocus><br>
+                </div>
+            </div>
+            <div class="row">
+                <label for="emil" class="col-sm-2 col-form-label">password</label>
+                <div class="col-auto">
+                    <input type="text" class="form-control" name="password" id="password" placeholder="Required"><br>
+                </div>
+                <div class="col-auto">
+                    <span id="PasswordHelpInline" class="form-text">
```
Comment (yammesicka) [inline app/templates/reset_password.html]: password-help-inline (also, you can use bootstrap's styling guide to add danger elements)

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/routers/reset_password.py (hunk 26 above)
Comment (kobyfogel) [reply app/routers/reset_password.py:82]: The token expiration time is 15 minutes. What would you suggest me to do then? that the main logic of this feature

File: app/internal/email.py [hunk 29]
```diff
@@ -177,3 +202,32 @@ def verify_email_pattern(email: str) -> bool:
...
+    :param user: ForgotPassword schema.
+        Contains user's email address, jwt verifying token.
+    :param background_tasks: (BackgroundTasks): Function from fastapi that lets
+            you apply tasks in the background.
+    returns True
+    """
+    params = f"?email_verification_token={user.email_verification_token}"
+    template = templates.get_template("reset_password_mail.html")
+    html = template.render(
+        recipient=user.username.strip("@"),
```
Comment (yammesicka) [inline app/internal/email.py]: `lstrip`

File: app/internal/security/ouath2.py [hunk 30]
```diff
@@ -30,17 +43,47 @@ def verify_password(plain_password: str, hashed_password: str) -> bool:
...
+    user: schema.ForgotPassword,
+    email: bool = False,
+) -> Union[schema.ForgotPassword, bool]:
+    """
+    Verifying database record by username.
+    Comparing given email to database record,
+    """
+    db_user = await User.get_by_username(
+        db=db,
+        username=user.username.strip("@"),
```
Comment (yammesicka) [inline app/internal/security/ouath2.py]: lstrip

File: app/routers/reset_password.py [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+        validated = False
+    if not validated:
+        return templates.TemplateResponse(
+            "reset_password.html",
+            {"request": request, "message": "Please check your credentials"},
+        )
+    await update_password(db, jwt_username, user.password)
+    message = "?message=Success reset password"
+    return RedirectResponse(
+        login_router.url_path_for("login_user_form") + f"{message}",
```
Comment (yammesicka) [inline app/routers/reset_password.py]: `str(message)`

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/templates/reset_password.html [hunk 32]
```diff
@@ -0,0 +1,44 @@
+{% extends "base.html" %}
+{% block content %}
+<div class="container">
+    <h5>Please choose a new password:</h4>
```
Comment (yammesicka) [inline app/templates/reset_password.html]: h4/h5

**Review by yammesicka — CHANGES_REQUESTED** [review]

**Pull request merged**